Gzip support

PhilipWhiteside opened the pull request:
Added GZIP support for package files, both automatic, and manual uploads. Default options if "0" (no GZIP). Option in settings to enable GZIP "1".

Resolves #403 

I have tested this with a few of my own devices. 

![Screenshot from 2021-11-09 13-30-09 148260240](https://user-images.githubusercontent.com/22427598/140940864-d6fd0cf8-90f1-490d-a50f-9ad775f2bbb3.png)

![Screenshot from 2021-11-09 13-04-53 001224612](https://user-images.githubusercontent.com/22427598/140941037-0e7e0e05-b73a-4ac1-8f25-23e0c20ed310.png)

ngalfas commented:
@PhilipWhiteside i just tested your code, works ok for automatic update.
I have the habbit of using the upload form for manual upload of my own firmwares.  This stills prohibits uploading of gziped file.
Can you check that ?

PhilipWhiteside commented:
Hey @ngalfas thanks for taking a look. Do you have the error description you saw?

It is working for me, screencast below. Are you enabling gzip in settings before uploading, or reloading the upload page after setting gzip to enabled?

![ta-working](https://user-images.githubusercontent.com/22427598/141614413-48888625-7426-4de0-9d92-51657be9118a.gif)

If gzip isn't enabled before the upload page, or the upload page is not refreshed, it will run before the gzip config option is saved.  Therefore it expects non-gzip files. I haven't altered any of the config save code, just added the additional option. If it's confusing being on the uploads page, I can remove it, so it has to be set in settings.

ngalfas commented:
This is so strange. I confirm i have your commit changes on my side. i saw the php changed lines on my side.
The checkbox is selected, i tried refreshing the form, still no good.

I even confirm the gzip setting is enabled on the session file.   (for a strange reason i see also some 0 byte session files created also)

So when i upload a .biz.gz file, i see that it erases previous files from /data/firmwares folder but it does not upload the .biz.gz files.

Error message in the upload.php page says 
`Minimal` Firmware: Wrong Format! Please upload Minimal Firmware BIN file! application/x-gzip!
Full Firmware: Wrong Format! Please upload Full Firmware BIN file! `application/x-gzip!`

PhilipWhiteside commented:
Thanks for that, I see the problem now. Not sure why it differs, but the difference is there. This is the current code

```php
if ($_FILES['minimal_firmware']["type"] == "application/octet-stream"
	|| $_FILES['minimal_firmware']["type"] == "application/macbinary"
	|| $_FILES['minimal_firmware']["type"] == "application/gzip") {
	if ($useGZIP == 1 || $useGZIP == "1") {
		$ext = "bin.gz";
	}
	else {
		$ext = "bin";
	}
}
else {
	throw new RuntimeException(
		__(
			"UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT",
			"DEVICE_UPDATE",
			$_FILES['minimal_firmware']["type"]
		)
	);
}
```

Specifically

```php
|| $_FILES['minimal_firmware']["type"] == "application/gzip") {
```

Where your file is coming up as `x-gzip`.

I will redo that part, add in `x-gzip`, and add in better error reporting based on whether gzip is enabled or not (Currently asking for `.bin`, not .`bin.gzip`.

PhilipWhiteside commented:
@ngalfas please can you try again with this latest commit? I have added the x-gzip as a supported format, and also added a gzip specific error output when gzip is enabled. 

New check:

```php
if ($useGZIP == 1 || $useGZIP == "1") {
	if ($_FILES['minimal_firmware']["type"] == "application/gzip"
		|| $_FILES['minimal_firmware']["type"] == "application/x-gzip") {
			$ext = "bin.gz";
	}
	else {
		throw new RuntimeException(
			__("UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT_GZIP", "DEVICE_UPDATE", $_FILES['minimal_firmware']["type"])
		);
	}
}
else {
	if ($_FILES['minimal_firmware']["type"] == "application/octet-stream"
		|| $_FILES['minimal_firmware']["type"] == "application/macbinary") {
			$ext = "bin";
	}
	else {
		throw new RuntimeException(
			__("UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT", "DEVICE_UPDATE", $_FILES['minimal_firmware']["type"])
		);
	}
}
```

New error messages:

```ini
UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT_GZIP = "Wrong Format! Please upload Minimal Firmware BIN.GZ file!  %s"
...
UPLOAD_FIRMWARE_FULL_WRONG_FORMAT_GZIP = "Wrong Format! Please upload Full Firmware BIN.GZ file!  %s"
```

ngalfas commented:
Alright :)
Now it's working as intended. This is a critical commit since many people are facing the same problem these days. Just check the latest issues..

Thanks for your effort.  
If you want to make it still better, i think that the upload form should be "autonomous" from the gzip config option.
You should check the files being uploaded INDIVIDUALLY. 
IE, right now you can't upload a .BIN minimal FW  and a .biz.gz full FW.  (this is exactly my habbit :) )
You can't upload "mixed" type of files.  
Again then, there can be a warning about this on upload form.

PhilipWhiteside commented:
That's a great idea! 

User can now upload either type. TA will detect it and act accordingly. This should work both ways around (mininal.bin.gz & full.bin; minimal.bin & full.bin.gz). I have partially tested this, as all my devices require gzip minimal and full, I can't test it fully. I do see the update URL matching the file name appropriately. Automatic updates are still consistent in file type, as per the config option setting by the user. 

Additionally, I worked out why the "use-gzip" setting for automatic wasn't applying the first time on upload, and have corrected this. No need to refresh to page. 

Thank you for your patience and guidance.

ngalfas commented:
Fantastic job. Now it's perfect, noone can go wrong with this.

Let's hope it goes quickly on a new version for everyone to update their installations.

thecode commented on tasmoadmin/lang/lang_en.ini:
```diff
@@ -270,7 +271,7 @@ UPLOAD_FIRMWARE_MINIMAL_ERR_NO_FILE = "Server Error: Minimal Firmware not upload
 UPLOAD_FIRMWARE_MINIMAL_UNKNOWN_ERROR = "Server Error: Minimal Firmware Unknown Error"
 UPLOAD_FIRMWARE_MINIMAL_TOO_BIG = "Minimal Firmware too big (max. %s)"
-UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT = "Wrong Format! Please upload Minimal Firmware BIN file!  %s"
+UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT = "Wrong Format! Please upload Minimal Firmware BIN or GIN.GZ file!  %s"
```
```suggestion
UPLOAD_FIRMWARE_MINIMAL_WRONG_FORMAT = "Wrong Format! Please upload Minimal Firmware BIN or BIN.GZ file!  %s"
```
Typo
I also think this should be lower case to match the released filename `bin` and `bin.gz`

PhilipWhiteside replied:
Great spot, compressed Gin, liquids can't be compressed unfortunately.

As for capitalisation, the original code-base uses capitalised BIN, presumably for emphasis. This PR sticks to convention already set by the project.

inverse commented:
@reloxx13 any change this could get merged + a new release? 

Let me know how I can help 🙏🏼

inverse commented on tasmoadmin/pages/upload.php:
```diff
@@ -149,6 +158,14 @@
 }
 elseif (isset($_REQUEST["auto"])) {
+	$useGZIP   = $Config->read("use_gzip_package");
+	if ($useGZIP == 1 || $useGZIP == "1") {
```
```suggestion
	if ($useGZIP == 1) {
```

Could be simplified to this since you're not using strict (`===`) equalities

PhilipWhiteside replied:
Looks great :-D

reloxx13 commented:
> @reloxx13 any change this could get merged + a new release?
> 
> Let me know how I can help 🙏🏼

thanks everyone.

currently the ci (travis) is broken and deprecated and has to be switched to github actions.   
i cannot build a new version without it.

inverse commented:
@reloxx13 I can prob help with the GH actions when I have time either today/tomorrow

mrneutron42 commented:
What version of TasmoAdmin will this be released in?
I'd like to see it get rolled into Home Assistant ASAP.

inverse commented:
> What version of TasmoAdmin will this be released in?
> I'd like to see it get rolled into Home Assistant ASAP.


It's pending the release automation to be fixed, there is a PR that could potentially solve it waiting for review.

marine1988 commented:
Why so many time to integrate this PR?

inverse commented:
> Why so many time to integrate this PR?

The release pipeline is broken on Travis and is being ported to GitHub actions. People that maintain this software do it in their spare time for mostly free.

#452 should help with the migration
